draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,69 +60,302 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+    uint32 min_batch_size;
+    select (query_type) {
+        case time-interval: Empty;
+        case fixed-size: uint32 max_batch_size;
+    }
+} QueryConfig;
 ~~~
 
+The purpose of `TaskConfig` is to include all parameters that are necessary
+for creating a new task in aggregators. It includes all the fields to be
+associated with a task (see task configuration in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.). Besides, `TaskConfig` also includes fields
+that useful for configuring a task in-band:
+
+* An opaque `task_info` that is specific to a task. For e.g. this can be a
+string describing the purpose of this task. It can also be the DAP task ID
+chosen by task author.
+
+* An opaque `vdaf_data` that contains any VDAF specific parameters for the
+chosen `vdaf_type`. The aggregators MUST pass `vdaf_data` to VDAF initialiser,
+based on the chosen `vdaf_type`.
+
+The definition of Time, Duration, Url, QueryType follow those in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.
+
+Note that the parameters that are not necessarily tied to a task (for e.g.
+collector hpke config), and secrets that should not be known by clients, like
+`vdaf_verify_key`, MUST still be exchanged out-of-band among aggregators.
 
 # Client Behavior
 
 > TODO(wangshan) Say how the Client constructs the extension body. Does the
 > extension change anything else about the upload flow?
 
+The client should know whether `task-prov` extension will be used and all
+parameters required for `TaskConfig` prior to constructing the extension
+body, either out-of-band from aggregators, or from information already saved on
+client.
+
+To use this extension, the client should first decide what is the task ID, see
+{{construct-task-id}}.
+
+## Construct task ID {#construct-task-id}
+
+For `task-prov` extension, a DAP task is not created before distributing task
+configuration to clients. Therefore, a DAP task ID may not be available to
+clients, aggregators and collector before uploading. Aggregator or collector
+MAY still choose a task ID for this task and deliver it to all clients along
+with `TaskConfig`. In this case, client MUST use that task ID. This arrangement
+SHOULD be agreed by aggregators, collector and clients out-of-band.
+
+Alternatively, task ID can be constructed from `task_info` and any other part
+of task configuration, as long as the generated task ID is determinstic and
+stay consistent across all parties. For example, the task ID can be a SHA256
+hash of the entire serialized `TaskConfig`. In this case, the `task_info`
+serves as an unique byte array to distinguish different tasks that share the
+same task parameters. Task ID can be constructed on server side and then
+distributed to clients, or constructed independently on clients and servers.
+When constructed independently, the mechanism used for creating the task ID
+must be known to both clients and the collector.
+
+> OPEN ISSUE: Should task ID construction from TaskConfig be enforeced?
```

Comment:
I think so. If the goal of this extension is to do as much in-band provisioning as possible, then permitting flexibility here defeats the purpose because it requires the parties to agree, out-of-band, on the method for choosing the batch ID.

As has been pointed out, diversifying the task ID is important. One way to do this is to rely on the `task_info` to be unique across the entire deployment. Alternatively, we could add a nonce field to the TaskConfig, which the task author should fill with random bytes.